### DISCUSSION TITLE
account/verify_credential API 403에러가 발생합니다.

### DISCUSSION THREAD

**Discussion by tjrkdgnl** [solvedac/unofficial-documentation#39] [질답] [answered]
BOJ에서 로그인 후,  쿠키에 저장되는  SolvedacToken을 account/verify_credential API 파라미터로 담아서 전송하는데 403에러가 계속 발생하고 있습니다.  원인이 무엇일까요 ㅜㅜ

**Comment by RanolP** [comment, marked as answer]
사이트에서 fetch를 호출할 때 보안적 안전을 위해 `credentials: "omit"`으로 설정되는 모양입니다. 직접 코드를 짜서 `credentials: "include"`로 설정했을 때 정상적으로 요청이 전송되는 걸 확인했습니다.

**Reply by RanolP** [reply]
다만 이 경우, 입력하는 토큰과는 무관하게 현재 로그인한 쿠키가 전송되므로 주의가 필요합니다.

**Reply by tjrkdgnl** [reply]
API 호출 시  request authentication을 basic header를 설정해야 credential을 보낼 수 있는 것으로 이해했는데 그럼 이때 사용되는 user /pwd는 BOJ의 아이디 비밀번호가 맞을까요?

**Reply by RanolP** [reply]
로그인 과정은 백준 SSO를 통해 수행되며, 이 과정에서 userId/password는 개입할 수 없을 거라 추측됩니다. solved.ac API는 `solvedacToken`이라는 이름의 쿠키에 JWT 토큰을 넣는 것을 동작하는 것으로 보이며. Playground에서 오류가 발생하던 건 보안 정책 문제라서 고쳤습니다.

**Reply by tjrkdgnl** [reply]
말씀은 이해했습니다! 하지만 제가 postman으로 API를 전송한 결과가 아래처럼 나오게 됩니다. token은 boj에서 로그인한 경우 쿠키에서 발췌한 것입니다.  

![image](https://user-images.githubusercontent.com/45396949/144733666-8f6c51df-8a83-490b-b064-87608b298f22.png)

**Reply by RanolP** [reply]
쿠키 위치가 아니라 쿼리 파라미터 위치에 넣으신 걸로 보입니다.

**Reply by tjrkdgnl** [reply]
제가 바보같은 실수를 했네요 ㅜㅜ header에 cookie를 셋팅하여 정상적으로 값을 받아오는 것을 확인했습니다!! 감사합니다.

**Reply by tjrkdgnl** [reply]
> 다만 이 경우, 입력하는 토큰과는 무관하게 현재 로그인한 쿠키가 전송되므로 주의가 필요합니다.

그런데 말씀하신 부분에 대해서 조금 걸립니다. 보안적인 문제를 말씀하신 것 같은데 제가 그럼 해당 token을 사용하고 전송할 때 어떠한 처리를 해야 안전하게 할 수 있을까요?

**Reply by RanolP** [reply]
현재 문서 사이트에서 관련 route를 테스트해볼 때 로그인된 계정의 토큰이 자동 입력됩니다. API Client를 직접 구현하시고 사용하시는 부분엔 따로 문제가 없을 겁니다.

**Reply by Yaminyam** [reply]
현재 사이트에서 token을 입력하지 않고 request를 보냈을때 로그인된 cookie값이 전달이 되지 않는것 같습니다

**Reply by RanolP** [reply]
제 환경(Windows 11 빌드 22000.652 / Firefox 101.0b4 (64-bit))에서는 재현이 되지 않습니다.
재현 가능한 조건을 제시해주시면 연구해볼 수는 있겠으나 해결을 확답드리기는 어렵습니다.
만약 재현 가능한 조건을 제시하시고자 하신다면 새로운 디스커젼이나 이슈를 만드시는 편이 좋겠습니다.
해결 전까지는 불편하시겠지만, solved.ac에서 쿠키 값을 추출해 테스트하시는 것을 제안드립니다.

**Reply by Yaminyam** [reply]
말씀해주신 firefox로 실행해보니 결과가 잘 출력됩니다 chrome환경에서 뭔가 다른 문제가 되는점이 있는것 같습니다

**Reply by Yaminyam** [reply]
edge, safari로 동일하게 시도해보았으나 firefox만 가지고있는 token을 가져와서 요청을 보내주는것 같습니다

**Reply by RanolP** [reply]
아무래도 계정 인증과 관련되어 있다보니 외부 사이트에서 접근하는 것이 힘든 것 같습니다.
테스트가 필요할 경우 브라우저가 제공하는 안전장치 바깥에서 테스트하시고, 여기서 제공되는 문서는 레퍼런스로만 삼는 것이 좋을 듯 싶습니다.

**Comment by tjrkdgnl** [comment]
확인했습니다. 감사합니다!!

2021년 12월 10일 (금) 오전 11:01, RanolP ***@***.***>님이 작성:
